django_ses/inbound.py

```diff
@@ -11,115 +13,129 @@
-
-        # Handle attachments
-        elif 'attachment' in content_disposition:
-            attachment = {
-                'filename': part.get_filename(),
-                'content_type': content_type,
-                'data': part.get_payload(decode=True)
-            }
-            attachments.append(attachment)
-
-    return plain_text, html_text, attachments
-
-
-def _get_common_headers(mail_obj):
-    common = mail_obj.get('commonHeaders', {})
-
-    return {
-        'subject': common.get('subject'),
-        'to': common.get('to'),
-        'message_id': common.get('messageId'),
-        'date': common.get('date'),
-    }
+class InboundEmail(TypedDict):
+    plain_text: str
+    html_text: str
+    attachments: array.array
+    subject: str
+    to: array.array
+    message_id: str
+    date: str
 
 
-def raw_handler(*args, **kwargs):
-    return kwargs
-
-
-def s3_handler(mail_obj, content, receipt, raw_message):
-    common_headers = _get_common_headers(mail_obj)
-    action = receipt.get('action', {})
-
-    action_type = action.get('type')
-    if action_type != 'S3':
-        logger.error(
-            f'Received action type ({action_type}) can\'t be handled by this '
-            f'handler')
-        raise UnprocessableError
-
-    bucket_name = action.get('bucketName')
-    object_key = action.get('objectKey')
-
-    s3_client = boto3.client(
-        's3',
-        aws_access_key_id=settings.ACCESS_KEY,
-        aws_secret_access_key=settings.SECRET_KEY,
-        aws_session_token=settings.SESSION_TOKEN,
-    )
-
-    response = s3_client.get_object(Bucket=bucket_name, Key=object_key)
-    content = response['Body'].read()
-
-    plain_text, html_text, attachments = parse_email(content)
-
-    return {
-        **common_headers,
-        'plain_text': plain_text,
-        'html_text': html_text,
-        'attachments': attachments,
-    }
-
-
-def sns_handler(mail_obj, content, receipt, raw_message):
-    common_headers = _get_common_headers(mail_obj)
-    action = receipt.get('action', {})
+class UnprocessableError(Exception):
+    pass
 
-    action_type = action.get('type')
-    encoding = action.get('encoding')
-    if action_type != 'SNS' or encoding != 'BASE64':
-        logger.error(
-            f'Received action type ({action_type}) or encoding ({encoding}) '
-            f'can\'t be handled by this handler')
-        raise UnprocessableError
 
-    try:
-        content = base64.b64decode(content, validate=True)
-    except binascii.Error:
-        logger.error(
-            'Content couldn\'t be decoded with base64. Invalid content?')
-        raise UnprocessableError
-
-    plain_text, html_text, attachments = parse_email(content)
-
-    return {
-        **common_headers,
-        'plain_text': plain_text,
-        'html_text': html_text,
-        'attachments': attachments,
-    }
+class BaseHandler:
+  def __init__(self, mail_obj, receipt, raw_message):
```

Comment:
Is it worth adding any typing to these arguments, or other methods on these handler classes?

Reply:
These 3 are the direct payload coming from SES. It might be useful to have types for these, but since all other signal handlers work the same way, I'd say this can be converted into another task/issue that covers all handlers. What you do think?

Reply:
Sure it can be a follow-up.